ccl/changefeedccl: TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart failed

ccl/changefeedccl.TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17708695?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17708695?buildTab=artifacts#/) on release-24.3 @ [e178e0dc95a2754ff7fecff7a589134380a8f164](https://github.com/cockroachdb/cockroach/commits/e178e0dc95a2754ff7fecff7a589134380a8f164):


```
=== RUN   TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/f236cfce7e32f4b492de4b4f1cd80c96/logTestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart3145899135
    test_log_scope.go:76: use -show-logs to present logs inline
    helpers_test.go:962: making kafka feed factory
    changefeed_test.go:6572: 
        	Error Trace:	pkg/ccl/changefeedccl/changefeed_test.go:6572
        	Error:      	Received unexpected error:
        	            	still waiting for job status; current running
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*jobFeed).WaitForStatus.func1
        	            	  | 	pkg/ccl/changefeedccl/testfeed_test.go:458
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	pkg/testutils/soon.go:71
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	pkg/util/retry/retry.go:213
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	pkg/testutils/soon.go:77
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoonError
        	            	  | 	pkg/testutils/soon.go:47
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*jobFeed).WaitForStatus
        	            	  | 	pkg/ccl/changefeedccl/testfeed_test.go:449
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart
        	            	  | 	pkg/ccl/changefeedccl/changefeed_test.go:6572
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1689
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_arm64.s:1222
        	            	Wraps: (2) still waiting for job status; current running
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f236cfce7e32f4b492de4b4f1cd80c96/logTestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart3145899135
--- FAIL: TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart (313.79s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #134903 ccl/changefeedccl: TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart failed [A-cdc C-test-failure O-robot T-cdc branch-master release-blocker]
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-44289

ccl/changefeedccl.TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart [failed](https://tanzanite.cluster.engflow.com/invocations/default/81e702c9-a29e-4aee-8d89-2a607de49e50?testReportRun=20&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on release-24.3 @ [bca9d708828f2e202df23b343efb8119a610c030](https://github.com/cockroachdb/cockroach/commits/bca9d708828f2e202df23b343efb8119a610c030):


```
=== RUN   TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart2174285561
    test_log_scope.go:76: use -show-logs to present logs inline
    helpers_test.go:962: making cloudstorage feed factory
    changefeed_test.go:6572: 
        	Error Trace:	pkg/ccl/changefeedccl/changefeed_test.go:6572
        	Error:      	Received unexpected error:
        	            	still waiting for job status; current running
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*jobFeed).WaitForStatus.func1
        	            	  | 	pkg/ccl/changefeedccl/testfeed_test.go:458
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	pkg/testutils/soon.go:71
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	pkg/util/retry/retry.go:213
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	pkg/testutils/soon.go:77
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoonError
        	            	  | 	pkg/testutils/soon.go:47
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*jobFeed).WaitForStatus
        	            	  | 	pkg/ccl/changefeedccl/testfeed_test.go:449
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart
        	            	  | 	pkg/ccl/changefeedccl/changefeed_test.go:6572
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1689
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (2) still waiting for job status; current running
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart2174285561
--- FAIL: TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart (144.69s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=20</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #134903 ccl/changefeedccl: TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart failed [A-cdc C-test-failure O-robot T-cdc branch-master release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/ccl/changefeedccl/cdctest/testfeed.go
+++ b/pkg/ccl/changefeedccl/cdctest/testfeed.go
@@ -7,6 +7,7 @@ package cdctest
 
 import (
 	"fmt"
+	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
@@ -75,6 +76,8 @@ type EnterpriseTestFeed interface {
 	Resume() error
 	// WaitForStatus waits for the provided func to return true, or returns an error.
 	WaitForStatus(func(s jobs.Status) bool) error
+	// WaitDurationForStatus waits for a specified time for the provided func to return true, or returns an error.
+	WaitDurationForStatus(dur time.Duration, statusPred func(status jobs.Status) bool) error
 	// FetchTerminalJobErr retrieves the error message from changefeed job.
 	FetchTerminalJobErr() error
 	// FetchRunningStatus retrieves running status from changefeed job.

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -6472,6 +6472,9 @@ func TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart(t *testing.T)
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
+	// Add verbose logging to help debug future failures.
+	require.NoError(t, log.SetVModule("changefeed_processors=1"))
+
 	// This test requires many range splits, which can be slow under certain test
 	// conditions. Skip potentially slow tests.
 	skip.UnderDeadlock(t)
@@ -6569,7 +6572,7 @@ func TestChangefeedTimelyResolvedTimestampUpdatePostRollingRestart(t *testing.T)
 	defer DiscardMessages(testFeed)()
 
 	// Ensure the changefeed is able to complete in a reasonable amount of time.
-	require.NoError(t, testFeed.(cdctest.EnterpriseTestFeed).WaitForStatus(func(s jobs.Status) bool {
+	require.NoError(t, testFeed.(cdctest.EnterpriseTestFeed).WaitDurationForStatus(5*time.Minute, func(s jobs.Status) bool {
 		return s == jobs.StatusSucceeded
 	}))
 }

--- a/pkg/ccl/changefeedccl/testfeed_test.go
+++ b/pkg/ccl/changefeedccl/testfeed_test.go
@@ -440,13 +440,15 @@ func (f *jobFeed) status() (status string, err error) {
 	return
 }
 
-func (f *jobFeed) WaitForStatus(statusPred func(status jobs.Status) bool) error {
+func (f *jobFeed) WaitDurationForStatus(
+	dur time.Duration, statusPred func(status jobs.Status) bool,
+) error {
 	if f.jobID == jobspb.InvalidJobID {
 		// Job may not have been started.
 		return nil
 	}
 	// Wait for the job status predicate to become true.
-	return testutils.SucceedsSoonError(func() error {
+	return testutils.SucceedsWithinError(func() error {
 		var status string
 		var err error
 		if status, err = f.status(); err != nil {
@@ -456,7 +458,11 @@ func (f *jobFeed) WaitForStatus(statusPred func(status jobs.Status) bool) error 
 			return nil
 		}
 		return errors.Newf("still waiting for job status; current %s", status)
-	})
+	}, dur)
+}
+
+func (f *jobFeed) WaitForStatus(statusPred func(status jobs.Status) bool) error {
+	return f.WaitDurationForStatus(testutils.SucceedsSoonDuration(), statusPred)
 }
 
 // Pause implements the TestFeed interface.

